### PR TITLE
API: createGroup & join group change

### PR DESCRIPTION
**Pull request by HugoGresse**
- update createGroup to provide a groupid and a password
- add joinGroup with user name and password to join an existing group

### CODE REVIEW

File: functions/src/api/v1/group/createGroup.ts [hunk 1]
```diff
@@ -47,12 +63,39 @@ export const createGroup = wrap(async (req: Request, res: Response) => {
...
+    }
+
+    let groupId: GroupId
+    if (requestedGroupId) {
+        try {
+            await GroupDao.get(requestedGroupId)
+            throw new GroupConflictError()
+        } catch (error) {
+            if (error instanceof GroupNotFoundError) {
+                await GroupDao.set(
```
Comment (MattiaPrimavera) [inline functions/src/api/v1/group/createGroup.ts:81]: why `set` ? Shouldn't `create` only do an `add` ?

File: functions/src/api/v1/group/createGroup.ts [hunk 2]
```diff
@@ -47,12 +63,39 @@ export const createGroup = wrap(async (req: Request, res: Response) => {
...
+                    members
+                )
+                groupId = requestedGroupId
+            } else {
+                throw error
+            }
+        }
+    } else {
+        groupId = await GroupDao.add(userId, name, password, members)
+        await GroupDao.update({
```
Comment (MattiaPrimavera) [inline functions/src/api/v1/group/createGroup.ts:95]: why doesn't `groupDao.add` set the `groupId` as well ?

File: functions/src/api/v1/group/joinGroup.ts [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+ *       400:
+ *         $ref: '#/components/responses/400'
+ *       401:
+ *         $ref: '#/components/responses/401'
+ *       403:
+ *         $ref: '#/components/responses/403'
+ *       412:
+ *         $ref: '#/components/responses/412'
+ */
+export const joinGroup = wrap(async (req: Request, res: Response) => {
```
Comment (MattiaPrimavera) [inline functions/src/api/v1/group/joinGroup.ts:42]: didn't we already have `/subscribe` endpoint for this ?

**Review by MattiaPrimavera — CHANGES_REQUESTED** [review]

File: functions/src/api/v1/group/createGroup.ts (hunk 1 above)
Comment (HugoGresse) [reply functions/src/api/v1/group/createGroup.ts:81]: here we want to add a group with a fixed id in advance, requiring set, not add.

File: functions/src/api/v1/group/createGroup.ts (hunk 2 above)
Comment (HugoGresse) [reply functions/src/api/v1/group/createGroup.ts:95]: the "add" add a new object within a collection, letting firestore pick the right id so we dont have any conflicted id.

File: functions/src/api/v1/group/joinGroup.ts (hunk 3 above)
Comment (HugoGresse) [reply functions/src/api/v1/group/joinGroup.ts:42]: /subscribe is only for the registerToken, but I'll add it optionnaly there too.

File: functions/src/api/v1/group/createGroup.ts (hunk 2 above)
Comment (MattiaPrimavera) [reply functions/src/api/v1/group/createGroup.ts:95]: (non blocking) but why calling `GroupDao.update` ? Isn't firebase id part of the entity, after the call to `add` ?

File: functions/src/api/v1/group/createGroup.ts (hunk 1 above)
Comment (MattiaPrimavera) [reply functions/src/api/v1/group/createGroup.ts:81]: (non blocking) I think `createGroup` logic may be expressed in a clearer way

**Review by MattiaPrimavera — APPROVED** [review]

File: functions/src/api/v1/group/createGroup.ts (hunk 2 above)
Comment (HugoGresse) [reply functions/src/api/v1/group/createGroup.ts:85]: the document id is not within the document itself on add, and is retrieve differently where quering the whole collection. It make things simpler to we know the id is always within the document.

File: functions/src/api/v1/group/createGroup.ts [hunk 4]
```diff
@@ -47,14 +64,49 @@ export const createGroup = wrap(async (req: Request, res: Response) => {
...
 })
+
+const tryCreateGroupWithPredefinedId = async (
+    groupId: GroupId,
+    userId: UID,
+    password: string,
+    members: Member[]
+): Promise<GroupId> => {
+    try {
+        await GroupDao.get(groupId)
```
Comment (ntchou) [inline functions/src/api/v1/group/createGroup.ts:102]: missing `return groupId`

Comment (HugoGresse) [reply functions/src/api/v1/group/createGroup.ts:102]: its below, we throw directly if the await is successful.

**Pull request merged**